### ISSUE TITLE
Think about integrating self-discovery into the protocol layer

### ISSUE THREAD

**Issue by Ingvord** [Ingvord/tango-v10-design-doc#2]
Currently a dedicated MySQL database is used to form device clusters. This can be replaced with self-discovery. Tango must benefit from P2P networking. This will reduce deployment and maintaining costs

**Comment by Ingvord** [comment]
This can be done using broadcasting. Issues to consider
1 - startup of a huge Tango system (20K devices broadcast at one time)
2 - multiple Tango hosts in a network

**Comment by andygotz** [comment]
Multiple networks and sub-networks going through multiple routers. Broadcasting does not usually go through routers unless it is configured to. Broadcasting is not a good idea. Something like DNS could be more scalable.

**Comment by Ingvord** [comment]
@andygotz , yeah configuring routers may be a great pain. I think we can support both methods. We need an automatic switch protocol from automatic discovery based on broadcasting to a tango host (aka DNS in this case) driven approach. This will also resolve the first issue from my previous comment - if there are over N devices registered -> switch to tango host. Still broadcasting may be very useful for small Tango installations i.e. beamlines and especially for getting started as it will smooth learning curve decreasing entering threshold. In case of broadcasting newcomers won't need to learn all things about Tango host, it will just work for them (I think this will remove the typical question on the forum "I have downloaded Tango, installed it and now Jive does not connect to the database").

**Comment by andygotz** [comment]
@Ingvord the pain comes from having tens and hundreds of routers to configure and the network guys not wanting to have broadcasts propagating. Concerning your use case for very small systems I could imagine some kind of auto-registration in the database or running without a database and printing out the port number to connect the device on. How to specify the device name?

**Comment by Ingvord** [comment]
@andygotz , auto-registration in the database is a good starting point. Once this is done we can think about broadcasting or multicasting (maybe these words are misleading due to their strict meaning in the network context). I would say we need a mechanism for a database to announce itself. Concerning the device name - I think it is responsibility of the device itself to know its name.

**Comment by andygotz** [comment]
Do you propose a new call for this? Today devices get their names from the database.

**Comment by Ingvord** [comment]
Concerning self discovery: we will drop this. It seems there is no easy way to implement it in a complex network. For this we will need some kind of a gossip protocol which is very tricky to implement correctly. On the other hand this does not give any considerable benefit. Self discovery is useful in a dynamic non-controlled environment e.g. multiple devices want to communicate and a random WiFi network. Since this is usually not the case for Tango i.e. Tango is naturally in controlled environment, because users that install Tango do it in there own environment. 

Still auto-registration is a must have.

@andygotz , yes I think we need such API.
